src/config.js

```diff
@@ -290,29 +290,17 @@ export default class Config {
       const preferredCacheFolder = opts.preferredCacheFolder || this.getOption('preferred-cache-folder', true);
 
       if (preferredCacheFolder) {
-        preferredCacheFolders = [preferredCacheFolder].concat(preferredCacheFolders);
+        preferredCacheFolders = [String(preferredCacheFolder)].concat(preferredCacheFolders);
       }
 
-      for (let t = 0; t < preferredCacheFolders.length && !cacheRootFolder; ++t) {
-        const tentativeCacheFolder = String(preferredCacheFolders[t]);
-
-        try {
-          await fs.mkdirp(tentativeCacheFolder);
-
-          const testFile = path.join(tentativeCacheFolder, 'testfile');
-
-          // fs.access is not enough, because the cache folder could actually be a file.
-          await fs.writeFile(testFile, 'content');
-          await fs.readFile(testFile);
-
-          cacheRootFolder = tentativeCacheFolder;
-        } catch (error) {
-          this.reporter.warn(this.reporter.lang('cacheFolderSkipped', tentativeCacheFolder));
-        }
+      const cacheFolderQuery = await fs.getFirstWriteableFolder(preferredCacheFolders);
+      for (const failedFolder of cacheFolderQuery.skipped) {
+        this.reporter.warn(this.reporter.lang('cacheFolderSkipped', failedFolder));
```

Comment:
`failedFolder` is not a string, so the display will probably be wrong. I think you meant `failedFolder.folder`? The variable name would probably be better as `skippedEntry` btw, so that it would be `skippedEntry.folder`

Reply:
Yup to both, good catch!